src/optimizer/rule/PushFilterDownLeftJoinRule.cpp

```diff
@@ -0,0 +1,129 @@
+    DCHECK_EQ(oldFilterNode->kind(), PlanNode::Kind::kFilter);
+    DCHECK_EQ(leftJoinNode->kind(), PlanNode::Kind::kLeftJoin);
+    auto* oldLeftJoinNode = static_cast<graph::LeftJoin*>(leftJoinNode);
+    const auto* condition = static_cast<graph::Filter*>(oldFilterNode)->condition();
+    DCHECK(condition);
+    const std::pair<std::string, int64_t>& leftVar = oldLeftJoinNode->leftVar();
+    auto symTable = octx->qctx()->symTable();
+    std::vector<std::string> leftVarColNames = symTable->getVar(leftVar.first)->colNames;
+    auto objPool = octx->qctx()->objPool();
+
+    // split the `condition` based on whether the varPropExpr comes from the left child
+    auto picker = [&leftVarColNames](const Expression* e) -> bool {
+        auto varProps = graph::ExpressionUtils::collectAll(e, {Expression::Kind::kVarProperty});
+        std::vector<std::string> propNames;
+        for (auto* expr : varProps) {
+            DCHECK(expr->kind() == Expression::Kind::kVarProperty);
+            propNames.emplace_back(*static_cast<const VariablePropertyExpression*>(expr)->prop());
+        }
+        for (auto prop : propNames) {
+            auto iter = std::find_if(leftVarColNames.begin(),
+                                     leftVarColNames.end(),
+                                     [&prop](std::string item) { return !item.compare(prop); });
+            if (iter == leftVarColNames.end()) {
+                return false;
+            }
+        }
+        return true;
+    };
+    std::unique_ptr<Expression> filterPicked;
+    std::unique_ptr<Expression> filterUnpicked;
+    graph::ExpressionUtils::splitFilter(condition, picker, &filterPicked, &filterUnpicked);
+
+    if (!filterPicked) {
+        return TransformResult::noTransform();
+    }
+
+    TransformResult result;
+    result.eraseAll = true;
+
+    // embed `filterPicked` into left Filter node
+    auto* newFilterNode = graph::Filter::make(
+        octx->qctx(),
+        const_cast<graph::PlanNode*>(oldLeftJoinNode->dep()),
+        objPool->add(
+            graph::ExpressionUtils::rewriteInnerVar(filterPicked.get(), leftVar.first).release()));
+    newFilterNode->setInputVar(leftVar.first);
+    newFilterNode->setColNames(leftVarColNames);
+    auto newFilterGroup = OptGroup::create(octx);
+    auto newFilterGroupNode = newFilterGroup->makeGroupNode(newFilterNode);
+    for (auto dep : leftJoinGroupNode->dependencies()) {
+        newFilterGroupNode->dependsOn(dep);
+    }
+
+    auto newFilterOutputVar = newFilterNode->outputVar();
+    if (filterUnpicked) {
+        // oldFilter and oldLeftJoin need not be reconstructed here
+        static_cast<graph::Filter*>(oldFilterNode)
+            ->setCondition(objPool->add(filterUnpicked.release()));
+        oldLeftJoinNode->setLeftVar({newFilterOutputVar, 0});
+        const_cast<OptGroupNode*>(leftJoinGroupNode)->setDeps({newFilterGroup});
+        result.newGroupNodes.emplace_back(const_cast<OptGroupNode*>(filterGroupNode));
```

Comment:
Generate new subplan rather than update this plan in-place.

Reply:
ACK.